### PR TITLE
Rewrite check-names.sh in python

### PR DESCRIPTION
**Pull request by dgreen-arm**
## Description
This PR aims to simplify check-names.sh by rewriting it and its sub-scripts into a single python script.
The reporting has also been extended to show which files invalid names are in. 

Note that there may be some issue with the Travis CI, however the main functionality of the script can be reviewed.

Internal ref: IOTCRYPT-229

### CODE REVIEW

File: tests/scripts/check-names.py
Comment (dgreen-arm) [reply tests/scripts/check-names.py]: Agreed, I'll change this to check for the Mbed TLS directories.

File: tests/scripts/check-names.py
Comment (dgreen-arm) [reply tests/scripts/check-names.py]: The code currently checks all subdirectories under the provided directories for .c and .h files, so this shouldn't be an issue as long as .h files remain in "include" and .c files in "library".

File: tests/scripts/check-names.py
Comment (dgreen-arm) [reply tests/scripts/check-names.py]: Its a simple enough change to ensure the leading character isn't a digit, so I'll change that.

File: tests/scripts/check-names.py
Comment (dgreen-arm) [reply tests/scripts/check-names.py]: `nm` was a previous dependency, however this wasn't immediately obvious as it was used in `list-symbols.sh`. Documenting the dependencies should be done though, so I'll add that.

The script should fail if nm isn't in the path, the output there looks like nm failed to read the object files. Can you give me the entire output of the script?

File: tests/scripts/check-names.py
Comment (dgreen-arm) [reply tests/scripts/check-names.py]: The script is expected to work on Linux and OS X. `nm` isn't available on Windows, so it won't work there.

File: tests/scripts/check-names.py
Comment (dgreen-arm) [reply tests/scripts/check-names.py]: This regex is more complicated than it needs to be, I'll change it to a simpler one with the same functionality.

I'm not using MBEDTLS because there was a typo of MBEDTSL that had slipped through.

File: tests/scripts/check-names.py
Comment (dgreen-arm) [reply tests/scripts/check-names.py]: Agreed, I'll make the change.

File: tests/scripts/check-names.py
Comment (mazimkhan) [reply tests/scripts/check-names.py]: Unless we want to tie this script to a specific Python version, we can just say ```#!/usr/bin/env python``` to allow working with whatever is default. That would need testing with both versions.

Comment (andresag01) [reply tests/scripts/check-names.py]: @mazimkhan: I disagree because I think this script was written for python 2, but if the default is python 3 then it probably wont work because these two versions are not compatible.

Comment (mazimkhan) [reply tests/scripts/check-names.py]: > That would need testing with both versions.

I did mention this.

Comment (dgreen-arm) [reply tests/scripts/check-names.py]: This was a mistake on my part, this should just be `python`. I was testing it with both versions and mistakenly left it as `python2`. The script is compatible with both python 2 & 3.

File: tests/scripts/check-names.py
Comment (dgreen-arm) [reply tests/scripts/check-names.py]: @andresag01, if you're still able to reproduce the same issue, could I get the entire script output?

Comment (andresag01) [reply tests/scripts/check-names.py]: @dgreen-arm: Sorry about the delay. I am still able to reproduce it, the following is the full output:
```
--> python ./tests/scripts/check-names.py
Analysing source code...
library/libmbedcrypto.a:md_wrapwarning: /Library/Developer/CommandLineTools/usr/bin/nm: no name list
warning: /Library/Developer/CommandLineTools/usr/bin/nm: no name list
1340 macros
147 enum-consts
1062 identifiers
862 exported-symbols
Names of identifiers: PASS
Names of actual-macros: PASS
Names of enum-consts: PASS
Names of identifiers: PASS
Likely typos: PASS
```
However, nm is actually in my path (please ignore my previous comment). I looked closely and nm is present, its just that the command the script issues (i.e. `nm -og library/libmbedcrypto.a library/libmbedtls.a library/libmbedx509.a`) is causing the warning `no name list`. Regardless of this, nm still prints out all the expected output and it does not return a failure code to the system, thus why the script continues to work as if nothing had happened...

Comment (dgreen-arm) [reply tests/scripts/check-names.py]: Thanks, I was wanting to check the counts of the macros etc. to see if they were affected by this. I'll see if I can work out what's actually happening here.

**Comment by gilles-peskine-arm** [comment]
The main goal of this rewrite was to get the script to explain problems better, but unfortunately this is not the case yet. After editing `aes.h` to remove the leading space from line 2 (a common issue after rewriting documentation), the output from the identifiers part is
```
Names of identifiers: FAIL
h - include/mbedtls/aes.h
```

1. Please make sure that all errors have an associated line number. If this is an error at all, it should say that it comes from line 2.
2. Please use a better parser for comments so that editing a comment does not cause this script to fail.

**Comment by andresag01** [comment]
@dgreen-arm, @Patater @sbutcher-arm: should we rework this PR sometime soon?

File: tests/scripts/check-names.py
Comment (simonbutcher) [reply tests/scripts/check-names.py:5]: Maybe not when the PR was written though.

Comment (mpg) [reply tests/scripts/check-names.py:5]: That's why I said "nowadays" :) But perhaps you're missing some context: we're likely to pick this up again, so my comment was mainly for the person who picks it up, as a reminder to update that line.

Comment (simonbutcher) [reply tests/scripts/check-names.py:5]: Just teasing. ;)

**Comment by yutotakano** [comment]
I think it's safe to say support for Python 2 is no longer necessary, and the minimum Python version would be 3.6 as per the MbedTLS readme. I'm reworking this PR with that in mind.

**Comment by mpg** [comment]
To expand slightly on what Gilles said: the versions documented in the Readme are the minimum required as of Mbed TLS 3.0, but other branches may have other requirements. Unfortunately those are not precisely documented, so more guesswork is involved to determine what's an acceptable requirement there, but the point I wanted to make is that `development`'s Readme is not the whole story.

**Comment by yutotakano** [comment]
Ready for review. The CI fails because:
- There are existing typos, not caught for some reason by the previous script
- `list-identifiers.sh --internal` for API/ABI compatibility exclusion has been replaced with `list_internal_identifiers.py`

**Comment by mpg** [comment]
I had a quick look at the logs, and indeed those are obvious typos. (Love the new output, btw, much more helpful, thanks.) I believe they were not caught by the previous script because it was only looking for things that start with `MBEDTLS_` while the new script looks for all things that start with `MBED`. Apparently it was too optimistic to assume that we would type `MBEDTLS_` reliably ;)

Could you fix those typos? We need a passing CI to consider this PR for merging.

Same for the API-ABI checker, except it's a bit more complicated because the change is in the `mbedtls-test` repository. Though it could also be solved here by restoring `tests/scripts/list-identifiers.sh` as a simple wrapper that just calls the new python script.

**Comment by yutotakano** [comment]
Addressed the typos! Waiting for CI.

> restoring tests/scripts/list-identifiers.sh as a simple wrapper that just calls the new python script.

Since `list_internal_identifiers.py` was created specifically for this, I think changing the `mbedtls-test` repository is better than creating yet another wrapper around this existing wrapper. (The new check-names script does not create temporary files, so a separate Python script was introduced to use the same parser). Or does it take time to update the CI that it's better to add a makeshift solution here?

**Comment by yutotakano** [comment]
Ah, I think I see what you mean now. It's better to add a shell wrapper around `list_internal_identifiers.py`, since it would make all existing tests pass and merge without issues & the CI can be changed later at any point. No need to change two things at once. Ok!

**Comment by yutotakano** [comment]
Jenkins and Travis has passed, but it's not reflected on GitHub.

File: tests/scripts/check_names.py
Comment (mstarzyk-mobica) [inline tests/scripts/check_names.py:86]: This needs a comment. What exactly does this function return?

File: tests/scripts/check_names.py
Comment (mstarzyk-mobica) [inline tests/scripts/check_names.py:825]: the `quiet` argument is passed around, until it reaches the `Problem`s child instance.
This argument will never change it's value once set, there is no point in passing it around.
The `Problem` class instances should be already created with the correctly set `quiet` attribute. You could implement a factory pattern for that but it would require a bit more effort.

Alternatively, you could utilize the __class__ mechanism, which "sort of" works like a factory:
```
from abc import ABC, abstractmethod

class Problem(ABC):
    def __str__(self):
        if self.__class__.quiet:  # this is a bit hacky, but requires least effort for the current state of the code
           return self.quiet_output()
        return self.verbal_output()

    @abstractmethod
    def quiet_output(self):
        pass

    @abstractmethod
    def verbal_output(self):
        pass

    @classmethod
    def set_quiet(cls, quiet): # this is a bit hacky, but requires least effort for the current state of the code
        cls.quiet = quiet

class Child(Problem):
    def quiet_output(self):
        return "child quiet output"

    def verbal_output(self):
        return "child verbal output"

Problem.set_quiet(True) #modifies parent class
c_quiet = Child() 
print(c_quiet)

Problem.set_quiet(False) #modifies parent class
c_verbal = Child()
print(c_verbal)

parent = Problem() # raises error, can't implement abstract class (assuming that this is what you want)
```
This should allow you to create and use the objects deriving from `Problem` without needing to pass the `quiet` option around the functions.

File: tests/scripts/check_names.py
Comment (mstarzyk-mobica) [inline tests/scripts/check_names.py:168]: The structure `if self.quiet: ...` is repeating through all children of the `Problem` class.
Have a look at the proposed code commented on line 825.

**Review by mstarzyk-mobica — CHANGES_REQUESTED** [review]
I am not done with review yet, but I am submitting what I noticed so far.

**Review by mpg — CHANGES_REQUESTED** [review]
Looks pretty good to me, I only have mostly minor comments.

File: tests/scripts/check_names.py
Comment (yutotakano) [reply tests/scripts/check_names.py:69]: I originally had `line_no`, `(start_pos, end_pos)` as arguments and internal field, but at this point Pylint told me I had too many function arguments (the limit is 5, I think). I thought it should be consistent between arguments and the fields, but the legibility argument is very sound -- I'll change it to unpack the argument into separate fields.

File: tests/scripts/check_names.py
Comment (yutotakano) [reply tests/scripts/check_names.py:825]: Sounds great. Thanks, didn't know Python had proper support for abstract classes. I'll do this.

File: tests/scripts/check_names.py
Comment (yutotakano) [reply tests/scripts/check_names.py:487]: I opted to keep the previous line in memory if the line is being skipped due to a comment. This is because a block comment can appear e.g. inside of a function declaration, and parsing should just continue going as if there was nothing there.

Comment (mpg) [reply tests/scripts/check_names.py:487]: Good point. Marking this as resolved, then.

**Review by mstarzyk-mobica — CHANGES_REQUESTED** [review]
General remark.
Using regex for parsing code is hard to achieve and even harder to maintain.
My proposal is using the `pycparser` to parse the code for us. Have a look at https://github.com/ARMmbed/mbedtls/pull/4875

**Comment by mstarzyk-mobica** [comment]
> General remark.
> Using regex for parsing code is hard to achieve and even harder to maintain.
> My proposal is using the `pycparser` to parse the code for us. Have a look at #4875

As discussed on team sync:
* This PR should not introduce pycparser and go along with regex-based parsing
* New issue created for enhancement request #4876

File: tests/scripts/check_names.py
Comment (yutotakano) [reply tests/scripts/check_names.py]: I've found out there is a convenient function called `.span()` that returns `(.start(), .end())`. This makes things a bit cleaner, and the benefits outweigh the issues with disabling Pylint there, so I've gone ahead and added back the `line_no` argument. Please see the code if my description isn't too clear!

File: tests/scripts/check_names.py
Comment (yutotakano) [reply tests/scripts/check_names.py:833]: I realised I could achieve the same thing by using class variables (modifying the parent class definition), which has the same behaviour without the boilerplate of ABCs, so I've gone with that approach. Please take a look and confirm I haven't made a mistake in my understanding!

**Review by mpg — DISMISSED** [review]
Thanks for addressing my feedback. Looks good to me.

**Review by mpg — APPROVED** [review]
LGTM

File: tests/scripts/check_names.py
Comment (mstarzyk-mobica) [reply tests/scripts/check_names.py:95]: This is still not addressed - can user create instance of the `Problem`?
If not, prevent it.

File: tests/scripts/check_names.py
Comment (mstarzyk-mobica) [reply tests/scripts/check_names.py:834]: The proposed piece of code was ment to address both problems I mentioned:
- passing around the `quiet` variable
- copying the `if quiet: print("something") ... else: print("something else")` logic to overriden `__str__` methods.

You have fixed the first issue. The second still remains.

**Review by mstarzyk-mobica — CHANGES_REQUESTED** [review]

**Comment by yutotakano** [comment]
Mateusz had a really good point about the Problem class being instantiable when it shouldn't be -- I've used Abstract Base Classes to make it an abstract class following his suggestion. This ensures code safety when the script is extended with more checks in the future by another person.

Ready for review once again!

**Review by mstarzyk-mobica — DISMISSED** [review]
Thanks for following my suggestion :)

**Review by mpg — APPROVED** [review]
LGTM.

Thanks @mstarzyk-mobica for the suggestion to factor the quiet logic into the base class, I hadn't noticed the opportunity and the code indeed looks cleaner this way.

**Comment by mpg** [comment]
I've restarted the pr-merge job which was skipped last time.

**Comment by mpg** [comment]
@yutotakano Can you create backports for the `development_2.x` and `mbedtls-2.16` branches ? Since this PR has a fairly long history, rather that backport the complete history as we normally do, I suggest first making a single commit that just adds the new files (as they are in this branch) and removes the olds ones. Then review if adjustments are need for that specific branch (for example, internal headers are handled differently), add new commits on top of the first one for these.

Feel free to ask if you have any questions about the backports.

**Comment by yutotakano** [comment]
Ok! Backports might take a while since I'm busy with university for the coming week, but I'll be able to work on solving the merge conflict. I'll take the second approach.

**Comment by mpg** [comment]
Sure, no problem! We usually wait for backport to be ready before merging a PR, but in this case I think we'll make an exception and merge this PR as soon as conflicts are resolved - otherwise new conflicts will keep appearing while you work on the backports.

**Review by mpg — APPROVED** [review]
The merge looks good to me.

**Review by paul-elliott-arm — APPROVED** [review]
This LGTM

**Comment by gilles-peskine-arm** [comment]
Merging without waiting for backports to avoid new merge conflicts.

File: tests/scripts/check_names.py
Comment (gilles-peskine-arm) [inline tests/scripts/check_names.py:226]: Annoyingly, in Python, `**` traverses symlinks to directories, which can be inefficient and can even lead to an infinite loop. Generally you shouldn't use `**` with Python's `glob.glob` if there can be symbolic links to directories. (By the way, bash 2.x/3.x have the same defect but it was fixed in bash 4. `**` in ksh and zsh is fine.)

https://github.com/ARMmbed/mbedtls/pull/4984

**Comment by gilles-peskine-arm** [comment]
Since this needs to be backported before #4760 (or after, but that's more work), I went ahead and did the backport: https://github.com/ARMmbed/mbedtls/pull/4985

**Pull request merged**
